### PR TITLE
Inline nested serializers

### PR DESCRIPTION
**Pull request by beauby**
Allows people to do stuff like:
```ruby
class PostSerializer < ActiveModel::Serializer
  attributes :id, :name, :body
  has_many :comments do
    attributes :id, :body
    belongs_to :author do
      attributes :name
  end
end
```

I'd really love for people to submit tests to try and break the implementation.

ref #1190 #1157 #1113

### CODE REVIEW

File: lib/active_model/serializer.rb [hunk 1]
```diff
@@ -132,10 +136,9 @@ def initialize(object, options = {})
       self.scope = instance_options[:scope]
 
       scope_name = instance_options[:scope_name]
-      if scope_name && !respond_to?(scope_name)
-        self.class.class_eval do
-          define_method scope_name, lambda { scope }
-        end
+      return unless scope_name && !respond_to?(scope_name)
+      self.class.class_eval do
+        define_method scope_name, -> { scope }
```
Comment (beauby) [inline lib/active_model/serializer.rb]: Style fix – not required for this PR.

File: lib/active_model/serializer.rb [hunk 2]
```diff
@@ -152,10 +155,10 @@ def attributes(options = {})
         end
 
       attributes.each_with_object({}) do |name, hash|
-        unless self.class._fragmented
-          hash[name] = send(name)
-        else
+        if self.class._fragmented
           hash[name] = self.class._fragmented.public_send(name)
+        else
+          hash[name] = send(name)
```
Comment (beauby) [inline lib/active_model/serializer.rb]: Style fix – not required for this PR.

File: lib/active_model/serializer/associations.rb [hunk 3]
```diff
@@ -29,7 +30,7 @@ class << base
 
       module ClassMethods
         def inherited(base)
-          base._reflections = self._reflections.try(:dup) || []
+          base._reflections = _reflections.try(:dup) || []
```
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: Style fix – not required for this PR.

File: lib/active_model/serializer/associations.rb [hunk 4]
```diff
@@ -61,26 +62,41 @@ def belongs_to(name, options = {})
...
-        def associate(reflection)
+        def associate(reflection, &block)
           self._reflections = _reflections.dup
 
           define_method reflection.name do
             object.send reflection.name
           end unless method_defined?(reflection.name)
 
-          self._reflections << reflection
+          _reflections << reflection
```
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: Style fix – not required for this PR.

File: lib/active_model/serializer/associations.rb [hunk 5]
```diff
@@ -61,26 +62,41 @@ def belongs_to(name, options = {})
...
+          define_nested_serializer(reflection.name.to_s.singularize, &block) if block_given?
+        end
+
+        # Define a nested serializer
+        # @param [String] resource_name The name of the association
+        # @return [void]
+        #
+        # @api private
+        #
+        def define_nested_serializer(resource_name, &block)
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: nice!

**Comment by beauby** [comment]
As usual, tests are failing because ruby < 2.0.1 cannot handle namespaced constants in `const_get`. Will fix that soon.

**Comment by beauby** [comment]
Fixed \o/

**Comment by NullVoxPopuli** [comment]
:+1: merging unless there are objections

**Comment by beauby** [comment]
Wait, I have to update the changelot

**Comment by beauby** [comment]
I'll rebase this as well, but let's wait for @bf4 and @joaomdmoura's remarks on code clarity and maybe some more tests.

**Comment by NullVoxPopuli** [comment]
kk

**Comment by NullVoxPopuli** [comment]
I think could be a good test, to make sure namespacing holds
```
PostSerializer
  -> CommentsSerializer
     -> PostSerializer # different attributes than the first
```

**Comment by beauby** [comment]
@NullVoxPopuli This is done in [here](https://github.com/rails-api/active_model_serializers/pull/1193/files#diff-06cb55f5339a435e811f59e814a7cee5R272) (the global `ShareSerializer` has an extra `date` attribute).

File: lib/active_model/serializer/associations.rb [hunk 6]
```diff
@@ -39,8 +40,8 @@ def inherited(base)
         # @example
         #  has_many :comments, serializer: CommentSummarySerializer
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: can you add another @example here with the fancy new block syntax? :-)

File: lib/active_model/serializer/associations.rb [hunk 7]
```diff
@@ -50,8 +51,8 @@ def has_many(name, options = {})
         # @example
         #  belongs_to :author, serializer: AuthorSerializer
         #
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: same here

File: lib/active_model/serializer/associations.rb [hunk 8]
```diff
@@ -61,26 +62,41 @@ def belongs_to(name, options = {})
         # @example
         #  has_one :author, serializer: AuthorSerializer
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: and here

File: lib/active_model/serializer/associations.rb (hunk 4 above)
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: :+1:

File: lib/active_model/serializer/associations.rb [hunk 9]
```diff
@@ -61,26 +62,41 @@ def belongs_to(name, options = {})
...
-          self._reflections << reflection
+          _reflections << reflection
+
+          define_nested_serializer(reflection.name.to_s.singularize, &block) if block_given?
+        end
+
+        # Define a nested serializer
+        # @param [String] resource_name The name of the association
+        # @return [void]
+        #
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: can you give an example input, and then provide an example class that's generated by this? thanks!

File: lib/active_model/serializer/reflection.rb [hunk 10]
```diff
@@ -42,13 +42,13 @@ class Serializer
       def build_association(subject, parent_serializer_options)
         association_value = subject.send(name)
         reflection_options = options.dup
-        serializer_class = ActiveModel::Serializer.serializer_for(association_value, reflection_options)
+        serializer_class = subject.class.serializer_for(association_value, reflection_options)
```
Comment (NullVoxPopuli) [inline lib/active_model/serializer/reflection.rb]: why this change?

File: test/fixtures/poro.rb [hunk 11]
```diff
@@ -258,4 +260,30 @@ def maker
...
+
+  belongs_to :author do
+    attributes :id
+  end
+
+  has_many :shares do
+    attributes :id, :platform
+
+    belongs_to :author do
+      attributes :id, :name
```
Comment (NullVoxPopuli) [inline test/fixtures/poro.rb]: I guess, a note for later - without sideloading, it's possible a have a very redundantly bloated resulting json

File: test/adapter/json/nested_serializers_test.rb [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+            @tweet.shares = [@share1]
+            @share1.author = @author
+            @author.posts = []
+            @author.roles = []
+            @author.bio = nil
+          end
+
+          def test_nested_serializers
+            actual = ActiveModel::SerializableResource.new(@tweet, adapter: :json).serializable_hash
+            expected = { tweet: { id: 1, body: 'Tweet 1', date: 'Jan 15', author: { id: 1 }, shares: [{ id: 1, platform: 'facebook' }] } }
```
Comment (NullVoxPopuli) [inline test/adapter/json/nested_serializers_test.rb]: can this me multi-line formatted? with single line hashes, it can be hard to tell where some objects end (sometimes)

File: lib/active_model/serializer/reflection.rb (hunk 10 above)
Comment (beauby) [inline lib/active_model/serializer/reflection.rb]: We need the lookup to happen on the parent serializer class so that it can fetch the nested serializer in priority.

Comment (NullVoxPopuli) [inline lib/active_model/serializer/reflection.rb]: cool

File: lib/active_model/serializer.rb [hunk 13]
```diff
@@ -112,10 +112,21 @@ def self.digest_caller_file(caller_line)
...
+
+      chain
+    end
+
     def self.get_serializer_for(klass)
       serializers_cache.fetch_or_store(klass) do
-        serializer_class_name = "#{klass.name}Serializer"
-        serializer_class = serializer_class_name.safe_constantize
+        # NOTE(beauby): Once we drop 1.9.3 support, we can lazify the map for better perfs.
+        serializer_class = serializer_lookup_chain_for(klass).map(&:safe_constantize).find { |x| x }
```
Comment (beauby) [inline lib/active_model/serializer.rb]: This is the dough of this PR.

File: lib/active_model/serializer/array_serializer.rb [hunk 14]
```diff
@@ -10,10 +10,9 @@ class ArraySerializer
       def initialize(resources, options = {})
         @root = options[:root]
         @object = resources
+        parent_serializer = options.fetch(:_parent_serializer, ActiveModel::Serializer)
         @serializers = resources.map do |resource|
-          serializer_class = options.fetch(:serializer) do
-            ActiveModel::Serializer.serializer_for(resource)
-          end
+          serializer_class = options.fetch(:serializer) { parent_serializer.serializer_for(resource) }
```
Comment (beauby) [inline lib/active_model/serializer/array_serializer.rb]: This is part of the dough of this PR (the whole `ArraySerializer` situation adds an intermediate object, so we need to keep track of the parent serializer).

File: lib/active_model/serializer/associations.rb [hunk 15]
```diff
@@ -29,58 +29,97 @@ class << base
...
         #  has_many :comments, serializer: CommentSummarySerializer
+        #  has_many :comments do
+        #    attributes :id, :content
+        #  end
         #
-        def has_many(name, options = {})
-          associate HasManyReflection.new(name, options)
+        def has_many(name, options = {}, &block)
+          associate(HasManyReflection.new(name, options), &block)
         end
```
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: This is part of the dough.

File: lib/active_model/serializer/associations.rb [hunk 16]
```diff
@@ -29,58 +29,97 @@ class << base
...
         #  belongs_to :author, serializer: AuthorSerializer
+        #  belongs_to :author do
+        #    attributes :id, :name
+        #  end
         #
-        def belongs_to(name, options = {})
-          associate BelongsToReflection.new(name, options)
+        def belongs_to(name, options = {}, &block)
+          associate(BelongsToReflection.new(name, options), &block)
         end
```
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: This is part of the dough.

File: lib/active_model/serializer/associations.rb [hunk 17]
```diff
@@ -29,58 +29,97 @@ class << base
...
         #  has_one :author, serializer: AuthorSerializer
+        #  has_one :author do
+        #    attributes :id, :name
+        #  end
         #
-        def has_one(name, options = {})
-          associate HasOneReflection.new(name, options)
+        def has_one(name, options = {}, &block)
+          associate(HasOneReflection.new(name, options), &block)
         end
```
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: This is part of the dough.

File: lib/active_model/serializer/associations.rb [hunk 18]
```diff
@@ -29,58 +29,97 @@ class << base
...
 
           define_method reflection.name do
             object.send reflection.name
           end unless method_defined?(reflection.name)
 
-          self._reflections << reflection
+          _reflections << reflection
+
+          define_nested_serializer(reflection.name.to_s.singularize, &block) if block_given?
+        end
```
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: This is the dough.

File: lib/active_model/serializer/associations.rb [hunk 19]
```diff
@@ -29,58 +29,97 @@ class << base
...
+        #  end
+        #
+        # @api private
+        #
+        def define_nested_serializer(resource_name, &block)
+          serializer_name = "#{resource_name.camelize}Serializer"
+          serializer = Class.new(ActiveModel::Serializer)
+          serializer.class_eval(&block)
+          const_set(serializer_name, serializer)
         end
```
Comment (beauby) [inline lib/active_model/serializer/associations.rb]: This is the dough.

File: lib/active_model/serializer/reflection.rb (hunk 10 above)
Comment (beauby) [inline lib/active_model/serializer/reflection.rb]: This is the dough. We need the lookup to happen on the parent serializer class.

File: lib/active_model/serializer/reflection.rb [hunk 20]
```diff
@@ -62,11 +62,12 @@ def build_association(subject, parent_serializer_options)
 
       private
 
-      def serializer_options(parent_serializer_options, reflection_options)
+      def serializer_options(subject, parent_serializer_options, reflection_options)
         serializer = reflection_options.fetch(:serializer, nil)
 
         serializer_options = parent_serializer_options.except(:serializer)
         serializer_options[:serializer] = serializer if serializer
+        serializer_options[:_parent_serializer] = subject.class
```
Comment (beauby) [inline lib/active_model/serializer/reflection.rb]: This is required because of the `ArraySerializer` situation.

File: lib/active_model/serializer/associations.rb (hunk 19 above)
Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: was all this block adding stuff because of having 1.9.3 support?

Comment (beauby) [inline lib/active_model/serializer/associations.rb]: What do you mean?

Comment (NullVoxPopuli) [inline lib/active_model/serializer/associations.rb]: I think I thought I was viewing a recent commit where you fixed the build errors on pre ruby 2. 
ignore me. haha

**Comment by beauby** [comment]
Closing in favor of #1225 #1226.
